### PR TITLE
lightning: support Re/ReregisterMySQL by different tls name

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve?

Issue Number: https://github.com/pingcap/ticdc/issues/3511

Problem Summary:

* support Re/ReregisterMySQL by different tls name
* fix checkpoint waitgroup datarace

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
```
1. push this branch to pingcap/tidb@lightning-tls
2. update dep in ticdc to  pingcap/tidb@lightning-tls
3. running tls test in ci (multi lightning loader with tls)
```
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by Ehco1996** [comment]
could you please take another look for this pr ? thanks 🧡🧡🧡 @glorv @gozssky

File: br/pkg/lightning/config/config.go [hunk 1]
```diff
@@ -552,27 +552,40 @@ type Security struct {
...
 	// RedactInfoLog indicates that whether enabling redact log
 	RedactInfoLog bool `toml:"redact-info-log" json:"redact-info-log"`
+
+	// this is used to set tls config for lightning in DM
+	// DM may running many lightning instances at same time, so we need to set different tls config name for each lightning
+	TLSConfigName string `toml:"tls-config-name" json:"tls-config-name"`
 }
 
-// RegistersMySQL registers (or deregisters) the TLS config with name "cluster"
+// RegistersMySQL registers the TLS config with name "cluster" or security.TLSConfigName
```
Comment (glorv) [inline br/pkg/lightning/config/config.go:561]:
```suggestion
// RegisterMySQL registers the TLS config with name "cluster" or security.TLSConfigName
```

**Comment by glorv** [comment]
@Ehco1996 Could you please add the test result that this PR can solve the multi TLS conflict issue.

**Review by glorv — COMMENTED** [review]
rest LGTM after test result is added.

File: br/pkg/lightning/config/config.go [hunk 2]
```diff
@@ -552,27 +552,40 @@ type Security struct {
 	KeyPath  string `toml:"key-path" json:"key-path"`
 	// RedactInfoLog indicates that whether enabling redact log
 	RedactInfoLog bool `toml:"redact-info-log" json:"redact-info-log"`
+
+	// this is used to set tls config for lightning in DM
+	// DM may running many lightning instances at same time, so we need to set different tls config name for each lightning
+	TLSConfigName string `toml:"tls-config-name" json:"tls-config-name"`
```
Comment (lance6716) [inline br/pkg/lightning/config/config.go:558]: seems you didn't change the DSN (`param.Connect()`) to use this `TLSConfigName`

Comment (kennytm) [reply br/pkg/lightning/config/config.go:558]: is this config something that a user wants to change, or just to be automatically assigned by DM?

if latter i suggest hiding this config from TOML.

File: br/pkg/lightning/config/config.go [hunk 3]
```diff
@@ -552,27 +552,40 @@ type Security struct {
 	KeyPath  string `toml:"key-path" json:"key-path"`
 	// RedactInfoLog indicates that whether enabling redact log
 	RedactInfoLog bool `toml:"redact-info-log" json:"redact-info-log"`
+
+	// this is used to set tls config for lightning in DM
```
Comment (gozssky) [inline br/pkg/lightning/config/config.go:556]:
```suggestion
	// TLSConfigName is used to set tls config for lightning in DM
```

File: br/pkg/lightning/config/config.go [hunk 4]
```diff
@@ -552,27 +552,37 @@ type Security struct {
 	KeyPath  string `toml:"key-path" json:"key-path"`
 	// RedactInfoLog indicates that whether enabling redact log
 	RedactInfoLog bool `toml:"redact-info-log" json:"redact-info-log"`
+
+	// TLSConfigName is used to set tls config for lightning in DM, so we don't expose this field to user
+	// DM may running many lightning instances at same time, so we need to set different tls config name for each lightning
+	TLSConfigName string
```
Comment (kennytm) [inline br/pkg/lightning/config/config.go:558]: to explicitly opt-out you need

```suggestion
	TLSConfigName string `toml:"-" json:"-"`
```

**Review by kennytm — APPROVED** [review]

File: br/pkg/lightning/config/config.go [hunk 5]
```diff
@@ -552,27 +552,37 @@ type Security struct {
...
-	default:
-		gomysql.DeregisterTLSConfig("cluster")
+		_ = gomysql.RegisterTLSConfig(sec.TLSConfigName, tlsConfig)
 	}
 	return nil
 }
 
+// DeregisterMySQL deregisters the TLS config with security.TLSConfigName
+func (sec *Security) DeregisterMySQL() {
+	if sec == nil {
```
Comment (glorv) [inline br/pkg/lightning/config/config.go:580]:
```suggestion
	if sec == nil || len(sec.CAPath) == 0 {
```

**Review by glorv — APPROVED** [review]

**Comment by glorv** [comment]
/hold
Waiting for lightning in DM test result.

**Pull request merged**